lib/core/request.js

```diff
@@ -206,41 +247,63 @@ class Request {
       channels.bodySent.publish({ request: this })
     }
 
-    if (this[kHandler].onRequestSent) {
-      try {
-        return this[kHandler].onRequestSent()
-      } catch (err) {
-        this.abort(err)
-      }
+    try {
+      this[kHandler].onRequestEnd?.()
+      this[kHandler].onRequestSent?.()
+    } catch (err) {
+      this.abort(err)
     }
   }
 
   onConnect (abort) {
     assert(!this.aborted)
     assert(!this.completed)
+    assert(!this.controller)
 
     if (this.error) {
       abort(this.error)
     } else {
       this.abort = abort
-      return this[kHandler].onConnect(abort)
+      return this[kHandler].onConnect?.(abort)
     }
   }
 
-  onResponseStarted () {
-    return this[kHandler].onResponseStarted?.()
+  onResponseStarted (resume) {
+    assert(!this.aborted)
+    assert(!this.completed)
+
+    this.controller = new Controller(this.abort, resume)
+    try {
+      this[kHandler].onResponseStart?.(this.controller)
+
+      if (this[kHandler].onResponseStarted?.() === false) {
+        this.controller.pause()
+      }
+
+      return !this.controller.paused
+    } catch (err) {
+      this.abort(err)
+    }
   }
 
-  onHeaders (statusCode, headers, resume, statusText) {
+  onHeaders (headers, statusCode, statusText) {
     assert(!this.aborted)
     assert(!this.completed)
+    assert(this.controller)
 
     if (channels.headers.hasSubscribers) {
       channels.headers.publish({ request: this, response: { statusCode, headers, statusText } })
     }
 
     try {
-      return this[kHandler].onHeaders(statusCode, headers, resume, statusText)
+      this[kHandler].onResponseHeaders?.(parseHeaders(headers), statusCode, statusText)
+
+      const controller = this.controller
+      if (this[kHandler].onHeaders?.(statusCode, headers, () => controller.resume(), statusText || '') === false) {
+        this.controller.pause()
+      }
+
+      return !this.controller.paused
```

Comment:
The `onResponseHeaders` is meant to also return a `boolean`, or how does it decide wether to stop or not the `controller`?

Reply:
Not sure I understand the question.

Reply:
nvm, found my the answer, I overlooked the `return` statement 👍